### ISSUE TITLE
Can't set `baseUrl`

### ISSUE THREAD

**Issue by simonwiles** [CloudCannon/pagefind#129]
Thank you for Pagefind -- it's fast become far my favourite way to do search for static sites :)  Neither the content discovery nor the Svelte UI are flexible enough for my use cases, for the most part (appreciate it's early days with this stuff, of course), but the performance and especially the bandwidth usage of the chunked index approach is really compelling.

Because I'm doing indexing a little differently from the normal flow I've run into one issue with deployment, and it looks like it's probably a bug:

1. When using the Javascript API and attempting set the configuration as described [in the docs](https://pagefind.app/docs/search-config/#configuring-via-the-pagefind-api), it's impossible to set the `basePath` option properly[^1].
2. This is because the call to `const pagefind = await import("/_pagefind/pagefind.js");` initializes a `PagefindInstance` and [tries to fetch `${this.basePath}pagefind-entry.json`](https://github.com/CloudCannon/pagefind/blob/main/pagefind/src/output/stubs/search.js#L101) before `await pagefind.options({ ... })` can be called.
3. Essentially there's no way to set the `basePath` until after it's needed, so if the default resolution doesn't work the index cannot be loaded.

The resolution of `basePath` is quite hard to follow in the code, tbh (I presume this is to support the `mergeIndex` stuff?, but I've not looked into that).

I can work around this in a number of ways for my own purposes.  My current preferred solution is to export the `Pagefind` constructor in `search.js` and have it accept an optional `options` object such that a subsequent call to `pagefind.options()` is redundant.  I can't immediately find a solution that doesn't constitute a breaking change to the existing API, or I'd open a PR (and the whole "multiple indexes" thing may make this an over-simplistic solution for those use cases anyway).

@bglw -- what do you think about all this?

[^1]: The same issue applies also when using the Svelte UI, fwiw.  Note that the docs describe a `bundlePath` option instead, that doesn't actually exist -- https://pagefind.app/docs/search-config/#bundle-path.  I was about to open a PR against the docs to fix this but, as described above, I think there's a bigger problem that must be solved.

**Comment by bglw** [comment]
Hello! 👋 Lots of great feedback in here ❤️ 

`basePath` is indeed tricky, and my main regret with the current API design is that everything is instantiated on import rather than behind a `pagefind.init()` function, for the reasons you're encountering. In the current flow, it is indeed only possible to have your `pagefind.js` file living alongside the rest of the bundled files. As it stands, the `basePath` option really only exists for the index merging feature.

Your preferred solution is also my preferred solution, but I'd love to find a way to implement this without it being a breaking change. Some options come to mind:

- Have the `Pagefind` constructor look for a `window.pagefindOptions` object, which would let you set options pre-initialization. _(bit gross)_
- Output a new recommended file `pagefind-api.js` with the new structure, and silently deprecate the current file (but continue to output it)
- Change the current file to _not_ initialize on import and expose an `init()` function, but call the `init()` function first if any of the current exposed functions are called.
  - This one is a soft-breaking change, in that no existing integration would _break_, but they might be marginally slower if the Pagefind initialization is delayed. 

Curious to hear your thoughts on the above options, I think we can find something that's an okay option. 

Also, on the UI note, the `bundlePath` option is distinct and is used by the UI directly to locate the `pagefind.js` file itself, so the docs are correct in this case.

**Comment by simonwiles** [comment]
Thanks for your reply!

I'd lean towards outputting a new file and deprecating the existing one, for my part.  I don't see any guarantees of API stability in the repo or the docs, and certainly I was reading a release number of `0.x` as indicating that (and of course it will always be possible to pin the version in the `npx` invocation or similar -- that's what I'm doing now, since I've vendored and shimmed the `0.9.1` source files).


> Also, on the UI note, the bundlePath option is distinct and is used by the UI directly to locate the pagefind.js file itself, so the docs are correct in this case.

Ah, that's interesting, sorry.  I was puzzled because when I tried to use it like this (as implied [in the docs](https://pagefind.app/docs/search-config/#configuring-via-the-pagefind-api)):

```js
const pagefind = await import("/path/to/_pagefind/pagefind.js");
await pagefind.options({
  bundlePath: "/path/to/_pagefind/",
});
```
I got the following output:

![image](https://user-images.githubusercontent.com/96218/200025590-b84721cd-78dd-481e-954e-ad8858da30bc.png)
which led me astray.  Might be possible to make this clearer when the docs are updated following any changes resulting from this issue.

Thanks again :)

**Comment by paulstroot** [comment]
I just posted another issue related to how basePath is set. #168

**Comment by bglw** [comment]
Hey @simonwiles & @paulstroot ! 👋

Good news — this has landed in Pagefind v1.0.0! ✨

See the full release notes here: https://github.com/CloudCannon/pagefind/releases/tag/v1.0.0 💙

Pagefind now initializes when you call `pagefind.init()` or any search function, so you can now run `pagefind.options()` after importing and before searching 🙂

Let me know if this works as expected ❤️
